src/lightning_app/components/auto_scaler.py

```diff
@@ -0,0 +1,473 @@
+    ) -> None:
+        super().__init__()
+        self.num_replicas = 0
+        self._work_registry = {}
+
+        assert work_cls is not None or is_overridden("create_worker", self, AutoScaler)
+        self._work_cls = work_cls
+        self._input_schema = input_schema
+        self._output_schema = output_schema
+        self.autoscale_interval = autoscale_interval
+
+        if max_replicas < min_replicas:
+            raise ValueError("max_replicas must be less than or equal to min_replicas.")
+        self.max_replicas = max_replicas
+        self.min_replicas = min_replicas
+        self.downscale_threshold = downscale_threshold or min_replicas
+        self.upscale_threshold = upscale_threshold or min_replicas * max_batch_size
+        self._last_autoscale = time.time()
+
+        worker_url = worker_url or "api/predict"
+        self.load_balancer = _LoadBalancer(
+            input_schema=self._input_schema,
+            output_schema=self._output_schema,
+            worker_url=worker_url,
+            max_batch_size=max_batch_size,
+            timeout_batch=timeout_batch,
+            cache_calls=True,
+            parallel=True,
+        )
+        for _ in range(min_replicas):
+            work = self.create_worker()
+            self.add_work(work)
+
+        logger.info(
+            f"Initialized AutoScaler("
+            f"min_replicas={min_replicas}, "
+            f"max_replicas={max_replicas}, "
+            f"timeout_batch={timeout_batch}, "
+            f"max_batch_size={max_batch_size})"
+        )
+
+    @property
+    def workers(self) -> List[LightningWork]:
+        works = []
+        for i in range(self.num_replicas):
+            work = self.get_work(i)
+            works.append(work)
+        return works
+
+    def create_worker(self, *args, **kwargs) -> LightningWork:
+        """Override this hook to customise the work creation process."""
+        return self._work_cls()
+
+    def add_work(self, work) -> str:
+        work_attribute = uuid.uuid4().hex
+        work_attribute = f"worker_{self.num_replicas}_{str(work_attribute)}"
+        setattr(self, work_attribute, work)
+        self._work_registry[self.num_replicas] = work_attribute
+        self.num_replicas += 1
+        return work_attribute
+
+    def remove_work(self, index: int) -> str:
+        work_attribute = self._work_registry[index]
+        del self._work_registry[index]
+        work = getattr(self, work_attribute)
+        work.stop()
+        self.num_replicas -= 1
+        return work_attribute
+
+    def get_work(self, index: int) -> LightningWork:
+        work_attribute = self._work_registry[index]
+        work = getattr(self, work_attribute)
+        return work
+
+    def run(self):
+        if not self.load_balancer.is_running:
+            self.load_balancer.run()
+
+        for worker in self.workers:
+            worker.run()
+
+        if self.load_balancer.url:
+            self.autoscale()
```

Comment:
we might need to add a fake state update otherwise `self.autoscale` is not called